---
title: 'Selection: practical 1'
author: "Lewis Spurgin"
output:
  html_document:
    css: js/styles.css
---

<script src="js/hide.js"></script>


## Introduction
We are going to begin looking at variation *within* genomes in order to make some inferences about selection. We will learn how to use sliding windows, and explore how to plot and interpret sliding-window based statistics. Finally, we will learn how to annotate a VCF file and look at some of the cool things we can do if we have a reference genome.

## The data
We will be using the *Tribolium castaneum* genomes from the genetic diversity practical. Copy the selection folder from Dati and set your working directory there.

## Site and window nucleotide diversity
In the genetic diversity practical we calculated $\pi$ in non-overlapping 200kb windows, and compared distributions of windowed-averages across experimental populations. For visualisation of how variation occurs along genomes, we sometimes wish to calculate statistics in *sliding* windows. This is done very simply using VCFtools - here is the code to calculate.

```{r eval = F}

vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 10000 --window-pi-step 2000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB10k
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 10000 --window-pi-step 2000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA10k

```

The above calculates $\pi$ on chromosome 2 in 10kb windows, with the window moving along in 2kb steps - separately for each population. Based on the above, can you repeat this analysis and calculate $\pi$ in:

- 50kb windows with 10kb steps
- 200kb windows with 50kb steps.

<div class="fold s">
```{r eval = F}
#50kb windows
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 50000 --window-pi-step 10000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB50k
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 50000 --window-pi-step 10000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA50k


#200kb windows
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 200000 --window-pi-step 50000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB200k
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --window-pi 200000 --window-pi-step 50000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA200k

```
</div>

And that's it! Explore your outputs using the command line and check that everything makes sense. We will look at how to plot this data later.

## Tajima's D

We can also use VCFtools to calculate Tajima's *D*. Can you calculate this statistic for chromosome ChLG2 in 50kb windows? 

<div class="fold s">
```{r eval = F}

#200kb windows
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --TajimaD 50000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB50k
vcftools --vcf ../genetic_diversity/TriboliumSSGenomes.vcf --chr ChLG2 --TajimaD 50000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA50k

```
</div>

What is the difference between the way we have constructed the windows between this analysis and the sliding window analysis of $\pi$? Do you think this will cause any problems?

## SnpEff
We will now use this very cool piece of software to annotate the SNPs in out VCF file, using the *Tribolium* reference genome. To run SnpEff we need work within the directory in which the program is contained. Use `cd` to navigate to your own copy of the `SnpEff` directory, which is within the `selection` directory.

SnpEff takes a VCF file, and compares the positions against a reference genome database. It then extracts lots of information from the genome database (whether the SNP is in an intron, exon etc., whether it is a zero-fold/four-fold site). The program comes with lots of pre-loaded genome databases, or we can generate ourselves. So our first task is to find out if they have a *Tribolium castaneum* genome. Go the help page for SnpEff, here:

http://snpeff.sourceforge.net/SnpEff_manual.html

Now see if you can figure out how to use the command line to i) get the software to print out the help message, ii) print out a list of all available databases, iii) View the first 10 rows of the table of databases, and iv) search the database table for "Tribolium":

<div class="fold s">
```{r eval = F}
#Get help
java -jar snpEff.jar

#Get list of databases
java -jar snpEff.jar databases #Lots!

#Get first few rows
java -jar snpEff.jar databases | head -n 10

#Search for tribolium
java -jar snpEff.jar databases | grep -i Tribolium

```
</div>

Hopefully you have found that there is a *Tribolium castaneum* database, and that it is simply called "Tribolium_castaneum". Now that we have this information, we can annotate our VCF files, using the following command:

```{r eval = F}
java -Xmx4g -jar snpEff.jar -v Tribolium_castaneum ../../genetic_diversity/TriboliumSSGenomes.vcf > Tribolium.ann.vcf
```

We will take you through each argument here separately - all this information can be obtained from the SnpEff help page.

- java *calls up java!*
- -Xmx4g *allocates 4gb memory to java*
- -jar snpEff.jar *specifies the java file to be run*
- -v *verbose mode, so you can see what the software is doing*
- Tribolium_castaneum *the genome database*
- ../TriboliumSSGenomes.vcf *your vcf file (not in this directory, hence "..")*
- Tribolium.ann.vcf *the output name I have specified - this can be whatever you like*

Once the program has run, we can use the associated software SnpSift to filter our annotated VCF file based on a range of criteria. Take a look at the SnpSift manual:

http://snpeff.sourceforge.net/SnpSift.html

Can you use SnpSift to write separate VCF files containing:

- Only synonymous mutations
- Only non-synonymous (missense) mutations
- Only variants in the OPTIX gene

<div class="fold s">
```{r eval = FALSE}
java -jar SnpSift.jar filter "( ANN[*].EFFECT has 'synonymous_variant' )"   Tribolium.ann.vcf > Tribolium.ann.syn.vcf

java -jar SnpSift.jar filter "ANN[*].EFFECT has 'missense_variant'" Tribolium.ann.vcf > Tribolium.ann.nonsyn.vcf

java -jar SnpSift.jar filter "ANN[*].GENE = 'OPTIX'" Tribolium.ann.vcf > Tribolium.ann.optix.vcf

```
</div>

Now we have created these new VCF files we can use VCFtools to calculate diversity separately at sites with different functional properties. Navigate back to your selection directory, then, for each population, calculate:

- Window $\pi$ (200kb windows, non-overlapping for non-synonymous sites on chromosome ChLG2
- Window $\pi$ (200kb windows, non-overlapping) for synonymous sites on chromosome ChLG2
- Sliding window $\pi$(1kb windows, 200bp step) for all sites in the OPTIX gene

<div class="fold s">
```{r eval = FALSE}
#non-syn diversity
vcftools --vcf SnpEff/Tribolium.ann.nonsyn.vcf --chr ChLG2 --window-pi 200000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out Tribolium_nonsyn_MB200k
vcftools --vcf SnpEff/Tribolium.ann.nonsyn.vcf --chr ChLG2 --window-pi 200000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out Tribolium_nonsyn_FA200k

#syn diversity
vcftools --vcf SnpEff/Tribolium.ann.syn.vcf --chr ChLG2 --window-pi 200000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out Tribolium_syn_MB200k
vcftools --vcf SnpEff/Tribolium.ann.syn.vcf --chr ChLG2 --window-pi 200000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out Tribolium_syn_FA200k

#optix
vcftools --vcf SnpEff/Tribolium.ann.optix.vcf --chr ChLG2 --window-pi 1000 --window-pi-step 200 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out Tribolium_optix_MB1k
vcftools --vcf SnpEff/Tribolium.ann.optix.vcf --chr ChLG2 --window-pi 1000 --window-pi-step 200 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out Tribolium_optix_FA1k

```
</div>

Ok, now we are ready to download our data for plotting. You will need to copy the following files to a suitable directory on your server (e.g. session7_output) and copy the following files there (10 in total):

- 200kb $\pi$ for all markers on chromosome ChLG2 (2 files - one per population)
- 200kb Tajima's D for all markers on chromosome ChLG2 (2 files - one per population)
- 200kb $\pi$ for synonymous variants on chromosome ChLG2 (2 files - one per population)
- 200kb $\pi$ for non-synonymous variants on chromosome ChLG2 (2 files - one per population)
- 1kb $\pi$ for all variants in the OPTIX gene (2 files - one per population)

(HINT - use `*.pi` to copy all files ending in ".pi" from a particular folder.)

## Make some plots

This is the plot we will be aiming to make:

```{r,fig.width = 8,fig.height = 7,echo = F,message = F, warning = F}

setwd("~/Documents/Teaching/PopGenBerlinLGS/session7_output/")
temp1 <- read.table("TriboliumSSGenomesMB200k.windowed.pi",header = T)
temp2 <- read.table("TriboliumSSGenomesFA200k.windowed.pi",header = T)

pi200 <- rbind(temp1,temp2)
pi200$Pop <- c(rep("MB",nrow(temp1)),rep("FA",nrow(temp2)))

library(ggplot2)
library(Rmisc)

FigA <- ggplot(pi200,aes(x = BIN_START/1000, y = PI,col = Pop))+
  geom_line()+
  theme_bw()+
  scale_color_manual(values = c("grey",grey(0.2)))+
  xlab("Position (kb)")+
  ylab(expression(pi))+
  theme(legend.position = "none")+
  annotate("text",x = -100,y = 0.012,label = "A")


####################

temp1 <- read.table("Tribolium_nonsyn_MB200k.windowed.pi",header = T)
temp2 <- read.table("Tribolium_syn_MB200k.windowed.pi",header = T)

inboth <- intersect(paste0(temp1$CHROM,temp1$BIN_START),paste0(temp2$CHROM,temp2$BIN_START))
temp1 <- subset(temp1,paste0(CHROM,BIN_START) %in% inboth)
temp2 <- subset(temp2,paste0(CHROM,BIN_START) %in% inboth)

temp1$NS_pi <- temp1$PI
temp1$S_pi <- temp2$PI

FigB <- ggplot(temp1,aes(x = S_pi, y = NS_pi))+
  geom_point(col = grey(0.2))+
  theme_bw()+
  ylab(expression(paste(pi," (non-synonymous sites)")))+
  xlab(expression(paste(pi," (synonymous sites)")))+
  xlim(0,1.4e-03)+
  ylim(0,1.4e-03)+
  geom_abline(intercept = 0, slope = 1,lty = 2)+
  annotate("text",x = 0,y = 0.0014,label = "B")
####################

temp1 <- read.table("Tribolium_nonsyn_FA200k.windowed.pi",header = T)
temp2 <- read.table("Tribolium_syn_FA200k.windowed.pi",header = T)

inboth <- intersect(paste0(temp1$CHROM,temp1$BIN_START),paste0(temp2$CHROM,temp2$BIN_START))
temp1 <- subset(temp1,paste0(CHROM,BIN_START) %in% inboth)
temp2 <- subset(temp2,paste0(CHROM,BIN_START) %in% inboth)

temp1$NS_pi <- temp1$PI
temp1$S_pi <- temp2$PI

FigC <- ggplot(temp1,aes(x = S_pi, y = NS_pi))+
  geom_point(col = "grey")+
  theme_bw()+
  xlim(0,1.4e-03)+
  ylim(0,1.4e-03)+
  ylab(expression(paste(pi," (non-synonymous sites)")))+
  xlab(expression(paste(pi," (synonymous sites)")))+
  geom_abline(intercept = 0, slope = 1,lty = 2)+
  annotate("text",x = 0,y = 0.0014,label = "C")


####################

multiplot(FigA,FigB,FigC,layout = matrix(c(1,1,2,3),nrow = 2,byrow = T))

####################

```

**Figure 1** Genetic diversity along chromosome 2 in two experimentally evolved *T. castaneum* populations subjected to male-biased (black) and female-biased (grey) sex ratios. **A** Sliding window nucleotide diversity (200kb windows with a 50kb step) along the entire chromosome. **B,C** Nucleotide diversity at synonymous and non-synonymous sites in non-overlapping 200kb windows, plotted separately for the male-biased (**B**) and female-biased (**C**) populations. The dotted lines in **B** and **C** represent $y = x$.

You should be able to have a fairly good go at reproducing the above on your own by now. If you get stuck, follow along below.

The top plot largely uses code/concepts that we have already introduced, so do have a go at reproducing this yourself. The code is below, for when you want to check.

<div class="fold s">
```{R}
#Load ggplot2
library(ggplot2)

#Read in data
temp1 <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesMB200k.windowed.pi",header = T)
temp2 <-  read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesFA200k.windowed.pi",header = T)

#Combine the data and add a population ID column
pi200 <- rbind(temp1,temp2)
pi200$Pop <- c(rep("MB",nrow(temp1)),rep("FA",nrow(temp2)))

#Make the plot
FigA <- ggplot(pi200,aes(x = BIN_START/1000, y = PI,col = Pop))+
  geom_line()+
  theme_bw()+
  scale_color_manual(values = c("grey",grey(0.2)))+
  xlab("Position (kb)")+
  ylab(expression(pi))+
  theme(legend.position = "none")+
  annotate("text",x = -100,y = 0.012,label = "A")

FigA

```
</div>

The bottom two plots require a little more data cleaning before we can make the plot. Let's consider panel **B** first, which plots non-synonymous and synonymous $\pi$ for the MB population. We first need to read in the data and take a look at it:

<div class="fold s">
```{R}
nonsyn <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/Tribolium_nonsyn_MB200k.windowed.pi",header = T)
syn <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/Tribolium_syn_MB200k.windowed.pi",header = T)

str(syn)
str(nonsyn)
```
</div>

Our aim here is to plot synonymous and non-synonymous diversity *from the same windows*. So what we need is a single data frame with non-synonymous and synonymous nucleotide diversity as separate columns. However, we can see that we have different numbers of windows in the two sets of results Can you figure out where and why the two data frames differ? Use R to do this.

<div class="fold s">
```{R}
#Identify which rows are in syn but not in non-syn
which(!(syn$BIN_START %in% nonsyn$BIN_START))

#And vice versa
which(!(nonsyn$BIN_START %in% syn$BIN_START))
```
</div>

This tells VCFtools gave us 4 windows with a calculation for $\pi$ at synonymous sites but not non-synonymous sites, and three windows where $\pi$ was calculated at non-synonymous sites but not synonymous sites. This is likely to have occurred because there were no synonymous/non-synonymous variants within those particular windows. We may wish to explore why that might be the case, but for now we can simply exclude sites that do not have both synonymous and non-synonymous variants.


<div class="fold s">
```{R}
inboth <- intersect(syn$BIN_START,nonsyn$BIN_START)

syn <- subset(syn,BIN_START %in% inboth)
nonsyn <- subset(nonsyn,BIN_START %in% inboth)
```
</div>

Check they are the same length
<div class="fold s">
```{R}
#Check they are the same length
nrow(syn) == nrow(nonsyn)
```
</div>

Check the bins are in the same order
<div class="fold s">
```{R}
#Check the bins are in the same order
all(syn$BIN_START == nonsyn$BIN_START)
```
</div>

Now make a new data frame and take a look

<div class="fold s">
```{R}
dd <- data.frame(CHROM = syn$CHROM,
                 BIN_START = syn$BIN_START,
                 synPI = syn$PI,
                 nonsynPI = nonsyn$PI)

#Take a look
head(dd)
```
</div>

Now we are ready to make our plot. The only thing that is different here is the addition of a manually added line with a slope of 1 and intercept of zero. I have also specified axis limits so panels B and C end up being the same.

<div class="fold s">
```{R}
FigB <- ggplot(dd,aes(x = synPI, y = nonsynPI))+
  geom_point(col = grey(0.2))+
  theme_bw()+
  ylab(expression(paste(pi," (non-synonymous sites)")))+
  xlab(expression(paste(pi," (synonymous sites)")))+
  xlim(0,1.4e-03)+
  ylim(0,1.4e-03)+
  geom_abline(intercept = 0, slope = 1,lty = 2)+
  annotate("text",x = 0,y = 0.0014,label = "B")
FigB
```
</div>

Once we have done this we can go through exactly the same process for the FA population, and change the colour.

<div class="fold s">
```{R}

#Read in the data
nonsyn <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/Tribolium_nonsyn_FA200k.windowed.pi",header = T)
syn <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/Tribolium_syn_FA200k.windowed.pi",header = T)

#Get unique 
inboth <- intersect(syn$BIN_START,nonsyn$BIN_START)

syn <- subset(syn,BIN_START %in% inboth)
nonsyn <- subset(nonsyn,BIN_START %in% inboth)

dd <- data.frame(CHROM = syn$CHROM,
                 BIN_START = syn$BIN_START,
                 synPI = syn$PI,
                 nonsynPI = nonsyn$PI)

FigC <- ggplot(dd,aes(x = synPI, y = nonsynPI))+
  geom_point(col = "grey")+
  theme_bw()+
  ylab(expression(paste(pi," (non-synonymous sites)")))+
  xlab(expression(paste(pi," (synonymous sites)")))+
  xlim(0,1.4e-03)+
  ylim(0,1.4e-03)+
  geom_abline(intercept = 0, slope = 1,lty = 2)+
  annotate("text",x = 0,y = 0.0014,label = "C")
FigC
```
</div>

Now we just need to put the plots together using `multiplot()`, and we're done!

<div class="fold s">
```{r,fig.width = 8,fig.height = 7,message = F, warning = F}
library(Rmisc)

multiplot(FigA,FigB,FigC,layout = matrix(c(1,1,2,3),nrow = 2,byrow = T))


```
</div>

## Delving a bit further

Discuss the figure with some colleagues. How do you interpret the findings? Do you think there is good evidence for selection anywhere on this chromosome?

If you do identify regions showing unusual patterns of diversity, it is generally a good idea to follow them up a bit further. Let's take a closer look at the region with highest $\pi$ in our dataset, found at around position 8MB. One thing we can do is zoom in on the region, using smaller sliding windows. Let's take a look with the 50kb window data. 

<div class="fold s">
```{R, warning = F}

#Read in data
temp1 <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesMB50k.windowed.pi",header = T)
temp2 <-  read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesFA50k.windowed.pi",header = T)

#Combine the data and add a population ID column
pi50 <- rbind(temp1,temp2)
pi50$Pop <- c(rep("MB",nrow(temp1)),rep("FA",nrow(temp2)))

#Make the plot
FigA <- ggplot(pi50,aes(x = BIN_START/1000, y = PI,col = Pop))+
  geom_point()+
  xlim(7000,10000)

FigA

```
</div>

The region of high diversity in the FA population is still there. How about using 10kb windows?

<div class="fold s">
```{R, warning = F}

#Read in data
temp1 <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesMB10k.windowed.pi",header = T)
temp2 <-  read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesFA10k.windowed.pi",header = T)

#Combine the data and add a population ID column
pi10 <- rbind(temp1,temp2)
pi10$Pop <- c(rep("MB",nrow(temp1)),rep("FA",nrow(temp2)))

#Make the plot
Fig <- ggplot(pi10,aes(x = BIN_START/1000, y = PI,col = Pop))+
  geom_point()+
  xlim(8000,9000)

Fig

```
</div>

There is still something going on, but the pattern is much less clear - any idea what might be going on?

Another thing we can do is look at a range of statistics in our region of interest. We can take a look at Tajima's D for the FA population.
<div class="fold s">
```{R, warning = F}

#Read in data
dd <-  read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesFA50k.Tajima.D",header = T)

#Make a variable to colour the region of interest
dd$Region <- ifelse(dd$BIN_START %in% seq(8e06,9e06,5e04),"HD region","Not HD region")

#Make the plot
Fig <- ggplot(dd,aes(x = BIN_START/1000, y = TajimaD,col = Region))+
  geom_point()+
  scale_colour_manual(values = c("black","grey"))

Fig

```
</div>

So this region is characterised by high diversity and high Tajima's D in the FA, but not MB, population. What is the explanation for this? What are the problems with the approach we have taken?

Finally, we can take a look at what genes are in this region of the genome. Go to the *Tribolium* genome browser here:

https://www.ncbi.nlm.nih.gov/genome/gdv/browser/?context=genome&acc=GCF_000002335.3

Use the genome data viewer to zoom into the relevant region on chromosome Lg2. What genes are there? Are there good candidates? Is looking for genes in this way a good idea?

##If you are finished
Download the table of gene locations for *Tribolium* LG2, here:

https://www.ncbi.nlm.nih.gov/genome/proteins/216?genome_assembly_id=271841

Read this table into R (HINT - this may need cleaning first!), and use R to:

- pull out a list of genes within a region of high genomic diversity
- plot genomic diversity for a region, and map on some gene locations - like this:


```{R,echo = F,warning = F,message = F}
#Read in data
temp1 <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesMB10k.windowed.pi",header = T)
temp2 <-  read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/TriboliumSSGenomesFA10k.windowed.pi",header = T)

#Combine the data and add a population ID column
pi10 <- rbind(temp1,temp2)
pi10$Pop <- c(rep("MB",nrow(temp1)),rep("FA",nrow(temp2)))

#Make the plot
Fig <- ggplot(pi10,aes(x = BIN_START/1000, y = PI,col = Pop))+
  geom_line()+
  theme_bw()+
  xlim(8000,9000)+
  xlab("Genomic position (KB)")

genes <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session7_output/LG2genes.txt",header = T)
genes <- subset(genes,Start %in% (8.4e06:8.6e06))
genes <- subset(genes,!duplicated(GeneID))
genes$y1 <- rep(0.022,nrow(genes))

Fig +
  xlim(8400,8600)+
  geom_segment(data = genes,aes(x = Start/1000,xend = Stop/1000,y = y1,yend = y1,col = NULL),lwd = 2)

```

