---
title: 'Selection: practical 2'
author: "Lewis Spurgin"
output:
  html_document:
    css: js/styles.css
---

<script src="js/hide.js"></script>

```{r setup, echo = F}
knitr::opts_knit$set(root.dir = '~/OneDrive - University of East Anglia/Documents/Teaching/PopGenBerlinLGS/session8_output/')
```


## Introduction
In this practical we will look at two approaches for detecting *recent* selection among populations - one that requires discrete, predefined populations, and one that does not

## The data
We will be using the great tit data set from the population structure practical. You do not need to copy the file across, but you do need to copy the selection2 folder from the ubuntu Dati to your home directory, as follows:

```{r eval = F}

cd
cp -r Dati/selection2 .
cd selection2

```

In this session we are looking for loci under selection, so we do not want to filter based on Hardy-Weinberg, or to prune based on LD, as this would remove loads of interesting information. However, we still want to exclude related individuals. You can create such a file using the information you generated in the population structure practical folder using a single line of plink code. Write a new plink file to the selection folder. The file should contain only unrelated individuals, but all markers

<div class="fold s">
```{R, eval = F}
plink --bfile ../population_structure/GreatTits --keep ../population_structure/GreatTitsFilteredPruned.rel.id --make-bed --out GreatTitsUnrel
```
</div>

## Instructions

### Locus $F_{ST}$
We can calculate $F_{ST}$ very easily within plink - in fact we have already done this in the population structure practical, using a python script. But in that instance we were only interested in average $F_{ST}$ across pairs of populations. We are now interested in the $F_{ST}$ for each marker across all of our populations. Can you figure out how to calculate this?

<div class="fold s">
```{R, eval = F}
plink --bfile GreatTitsUnrel --family --fst --out GreatTitsUnrel
```
</div>

Check the plink help and make sure you are happy with what the "family" tag is doing. Regardless, is this the best way to define our populations? What did our Admixture results show? We will now try to use our Admixture results from the $K$ = 4 analysis to assign each individual to an "Admixture population", then run a locus-based $F_{ST}$ analysis across the Admixture populations. If you think you can do this, go for it! Otherwise, step-by-step instructions below.

Remember that your ".Q" file from the Admixture analysis has one line per individual, with individuals in the same order as in your plink file. It then has the proportion of each cluster assigned to each individual as a separate column (so 4 columns for the $K$ - 4 analysis). What we need to do is detene, for each individual, which is the 'largest' of the four clusters. We can do this in R. Start up R in your server console, then load your admixture output into a data frame.

<div class="fold s">
```{R, eval = F}
R
library(tidyverse)
dd <- read_delim("../population_structure/admixture/GreatTitsThinned.4.Q",delim = " ", col_names = F)
```
</div>

Change the column names to "K1" to "K4".

<div class="fold s">
```{R, eval = F}
colnames(dd) <- paste0("K",c(1:4))
```
</div>

We now wish to generate a vector that tells us, for each individual, which of K1 to K4 is largest. Can you figure out how to do this, either using a `for()` loop, or using `apply()`?

<div class="fold s">
```{R, eval = F}
#Using a for loop
output <- rep(NA,nrow(dd))
for(i in 1:nrow(dd))
{
  output[i] <- names(which.max(dd[i,]))
}

#Using apply - faster and neater
output <- apply(dd,1,function(x) names(which.max(x)))

```
</div>

Make sure that you fully understand how the above code has worked. Look up the help files for each function in the code, and discuss with your colleagues. 

To define populations in plink, we are going to need to give it a file with our original population IDs in the first column, individual IDs in the second, and new population IDs in the third. We can create this using our output, and our.fam file. Load the .fam file into R

<div class="fold s">
```{r, eval = F}
pops <- read_delim("GreatTitsUnrel.fam", delim = " ",col_names = FALSE)
```
</div>

Now write an output file.

<div class="fold s">
```{r, eval = F}
towrite <- data.frame(oldpop = pops$X1,ind = pops$X2,newpop = output)
write_delim(towrite,"K4pops.txt",delim = "\t")
```
</div>

Now we can quit R, and use plink to calculate $F_{ST}$ based on our admixture proportions.

<div class="fold s">
```{r, eval = F}
plink --bfile GreatTitsUnrel --within K4pops.txt --fst --out GreatTitsK4
```
</div>

### EigenGWAS

Have a read of the EigenGWAS documentation here. This approach, along with a few other closely related approaches uses a GWAS on eigenvectors from a PCA to detect loci under divergent selection.

Navigate into the EigenGWAS folder. Now try to figure out how to run EigenGWAS on the first four eigenvectors from a PCA.

<div class="fold s">
```{r eval = F}
java -jar gear.jar eigengwas --bfile ../GreatTitsUnrel --ev 4 --out GreatTitsUnrel
```
</div>

This analysis will take a little while to run - while it's running, you can start downloading your other files - a list of what you need to download is below.

Given that EigenGWAS is a new program, it would be sensible to check the output files from the EigenGWAS carefully. 

### Making plots
Copy the following files to a folder called session8_output:

- $F_{ST}$ between your geographic populations (first plink analysis, ending in .fst)
- $F_{ST}$ between your admixture populations (second plink analysis, ending in .fst)
- Your Eigenvectors from the EigenGWAS analysis (ending in .eigenvec)
- Your EigenGWAS results (four files, all ending in .egwas)

Now download this file to your laptop using scp or pscp.

We are going to do the plotting slightly differently this time. I will introduce you to some of the plots and functions you can run, then it is up to you to make a publication quality figure from our selection analyses.

Let's first explore our $F_{ST}$ analyses.


<div class="fold s">

```{r message = F}
library(tidyverse)
library(cowplot) #for later
pop_fst <- read_delim("GreatTitsUnrel.fst", delim = "\t")
pop_fst
```


</div>

We can explore the distribution of $F_{ST}$ values using `geom_histogram()` or `geom_density()`.

<div class="fold s">
```{r,fig.width=7,fig.height=3,warning = F,message = F}

h1 <- ggplot(pop_fst,aes(x = FST))+
  geom_histogram()
d1 <- ggplot(pop_fst,aes(x = FST))+
  geom_density()

plot_grid(h1,d1)
```
</div>

What do these distributions tell you about drift and selection in this system?

We can also read in our $F_{ST}$ values from our admixture populations, in order to compare marker-based $F_{ST}$ values.




<div class="fold s">

```{r, message = F}
ad_fst <- read_delim("GreatTitsK4.fst", delim = "\t")
ad_fst
```

</div>


Think of a good way of plotting the two sets of $F_{ST}$ values against one another in a way that summarises the data.

<div class="fold s">

```{r,fig.width=6,fig.height=5,message = F}

pop_fst$ADFST <- ad_fst$FST

ggplot(pop_fst,aes(x = FST,y = ADFST))+
  geom_point(alpha = 0.3) +
  theme_bw()

```
</div>


Check the correlation coefficient between the two measures...

<div class="fold s">
```{r}
cor(pop_fst$FST,pop_fst$ADFST)

#Need to remove NAs and try again
pop_fst <- filter(pop_fst,!is.na(FST) & !is.na(ADFST))
cor(pop_fst$FST,pop_fst$ADFST)
```
</div>

Let's now take a look at the EigenGWAS data. First read in the eigenvector file and look at the data...

<div class="fold s">
```{r, message = F}
ev <- read_delim("GreatTitsUnrel.eigenvec", delim = "\t", col_names = F)
ev
```
</div>



This conveniently has the population and individual IDs as the first two columns. Define come column names.

<div class="fold s">
```{r}
colnames(ev) <- c("Population","Individual",paste0("PC",c(1:4)))
```
</div>

Now make some PCA plots

<div class="fold s">
```{r,fig.width=10,fig.height=3,message = F}
F1 <- ggplot(ev,aes(x = PC1,y = PC2,col = Population))+
  geom_point()

F2 <- ggplot(ev,aes(x = PC3,y = PC4,col = Population))+
  geom_point()

plot_grid(F1,F2)
```
</div>

EigenGWAS runs a GWAS on each of these four Eigenvectors. What do these four analyses tell us about selection?

Let's load in the analysis from the first Eigenvector and take a look at the results.

```{r, message = F}
eg1 <- read_delim("GreatTitsUnrel.1.egwas", delim = "\t")
eg1
```


Let's see how well out EigenGWAS p value and $F_{ST}$ are correlated
```{r}
cor(eg1$PGC,eg1$Fst)

#Need to remove NAs and try again
eg1 <- eg1[complete.cases(eg1),]
cor(eg1$PGC,eg1$Fst)

```

They are strongly negatively correlated. Is that what we expect?

Finally, we will use ggplot to produce a basic Manhattan plot of the corrected EigenGWAS p-values. Try produce the plot below.

<div class="fold s">
```{r,warning = F, message = F, fig.width=10,fig.height = 4}
eg1 %>%
  mutate(logp = -log10(PGC),
         rank = c(1:nrow(eg1))) %>%
  ggplot(aes(x = rank,y = logp,col = factor(CHR)))+
  geom_point()+
  theme_bw()+
  theme(legend.position = "none")+
  scale_color_manual(values = rep(c("navy","orange"),20))

```
</div>



## Over to you
Make a publication-quality figure summarising the selection analyses you have done. You can adapt figures we have already made or (better still) make completely new ones. For example, you may want to consider:

- Manhattan plots for the other eigenvectors
- Manhattan plots for FST
- qqplots from the eigenGWAS analysis
- Going to the great tit reference genome (link below) and finding whether SNPs under divergent selection are in genes
  https://www.ncbi.nlm.nih.gov/genome/gdv/browser/?context=genome&acc=GCF_001522545.2
- Combining your selection and population structure plots into a single figure.

Be creative!

