---
title: 'Introduction to R: practical'
author: "Lewis Spurgin"
output:
  html_document:
    css: js/styles.css
---

<script src="js/hide.js"></script>


## Introduction
The aim of this practical is to get you comfortable using R and R studio. By the end of this practical, you should be able to explore data frames, write a short program to perform some basic data processing, and plot and test some results. We will also have our first go at reading data into R. Some general tips:

- Write all your commands in a script and run them from there
- Remove commands from your script that are not a useful part of your analysis
- Annotate your script well (using `#`). No amount of annotation is too much if it helps you understand what you're doing
- Use R's help (e.g. `?mean`) if you are not sure what a function does
- Google is your friend

All the code you will need for this practical is included in this handout. However, you should try to figure things out for yourself before clicking on "show".

## The data
I have simulated some genetic data from a bird species captured from mainland and island populations. All simulated samples have been genotyped at five nucleotide polymorphism (SNP) markers, at each of which two alleles are found. We wish to know whether, at these loci, the mainland and island populations have different levels of genetic diversity.

## Instructions
Our aim is to calculate *observed heterozygosity* (i.e. the proportion of heterozygote individuals) at each locus, separately for the two populations in our dataset. We will plot these proportions, and test whether levels of heterozygosity differ between the two populations. 

*If you think you can do this without guidance, please go ahead and do so*. Otherwise, step-by-step instructions are provided below.

### Download, read in and explore the data.
On the amazon server, copy the "intro_to_r" folder from my directory to your home directory as follows:

```{r eval = F}
cd
cp -r Dati/intro_to_r .
```

Now, open a fresh terminal or command prompt on your laptop, that is not logged onto the server. Using `scp` (mac/linux) or `pscp` (windows), download the "intro_to_r" folder and all its containing files from your home directory on the amazon server to your laptop.

Now open up your file explorer. Find the "intro_to_r" folder you have just downloaded, and open the file "Mainland_Island_SNP_data.csv". You can open this in a spreadsheet editor (e.g. excel), or in your text editor. You should be able to see that the data has 12 columns, and lots of rows. Each row is an individual. The columns correspond to individual ID, population ID, followed one column per allele at each of the five SNP loci. This is a standard way of storing SNP genotype data, albeit one of many.

Now open R studio. Open a new script using *File > New File > R Script*. Set your working directory to the "intro_to_r" folder, by clicking *Session > Set Working Directory > Choose Directory*, or using the shortcut `ctrl+shift+h`.

We are going to use the tidyverse library to read in and manipulate our data. We need to load in this library, which you have hopefully already installed (see the software instruction sheet we sent before the course if not). At the top of our script, type the following:

```{r, echo = T,message=FALSE}
library(tidyverse)
```

Click on the line in the script and either click "Run", or press `ctrl+enter` to execute the line of code. Hopefully you get a message showing the packages being attached.

Now, in the next line of your script type in the commands to read in the *Mainland_Island_SNP_data.csv* file, and assign the dataframe to a suitable object name. We will use tidyverse's `read_csv()` instead of the traditional `read.csv()` command (**note that we are using an underscore and not a full stop**):

```{r, message = F}
dd <- read_csv("Mainland_Island_SNP_data.csv")
```


Click on the "environment" tab in the top-right panel of R studio. The object `dd` should now be listed. If so, the data has been successfully loaded into R. 

We can call up the tibble `dd` by just typing its name into the console or script and executing. You should get something like the following:

```{r}
dd
```

Tibbles are, by default, printed in a user-friendly way. We do not get lots of information dumped to our screen, and R includes lots of useful information for us. From the output above, you should be able to answer the following:

- How many individuals are there in your dataset?
- How many populations are there?
- How has R stored the SNP allele data?

We can call up individual variables, and perform functions on them, using the `$` notation. Note that, however, R will dump the entire variable to the screen, so we should be careful if we have very large datasets.

```{r}
dd$Pop
```

We can also use square bracket notation to call up individual rows and columns. Using this notation, get R to print out the following (if you can't remember how to do this, use the internet):

1. The 45th row

<div class="fold s">
```{r}
dd[45,]
```
</div>

2. The 2nd column

<div class="fold s">
```{r}
dd[,2]
```
</div>

3. The genotypes for the 22nd individual

<div class="fold s">
```{r}
dd[22,3:ncol(dd)]
#For the more experienced - can you do think of a better way to do this using grep()?
```
</div>


###Define our genotypes
In our dataset we have separate information for each of the two alleles found in each individual. For example, at our first SNP locus we can see from when we ran the `head()` function that the first individual in our dataset has "`r dd$SNP_1.1[1]`" for its first allele, and "`r dd$SNP_1.2[1]`" for its second allele. This individual is therefore a `r ifelse(dd$SNP_1.1[1] == dd$SNP_1.2[1],"homozygote","heterozygote")` at this locus. We can tell R to use this information to tell us about heterozygosity. But first, we need to get an understanding of logical operators in R. Try typing the following commands directly into your console in R studio:

```{r,eval = F}
4 < 3
4 > 3

4 = 3 #gives us an error. Why?
4 == 3 

x = 4
x == 4

"cat" == "dog"

x <- "cat"
x == "cat"
```

Hopefully you have figured out that a single equals sign and double equals sign mean very different things to R. A single equals sign generally works in the same way as the `<-` operator (with a few subtle differences), whereas a double equals sign asks a logical question, and returns a value of `TRUE` or `FALSE`.

So coming back to our SNP data, we can use the `==` operator to ask whether the first individual in our dataset is a homozygote at the first locus, by asking whether the variables `SNP_1.1` and `SNP_1.2` for this individual are the same.

<div class="fold s">
```{r}
# Get the data for the first individual
ind1 <- dd[1,]

# Now ask whether Allele 1 and Allele 2 are identical
ind1$SNP_1.1 == ind1$SNP_1.2

```
</div>

Because R works so nicely with vectors, we can do this for every individual in our dataset. Can you generate a list of `TRUE` and `FALSE` values telling you whether each individual is a homozygote at the `SNP_1` locus?

<div class="fold s">
```{r}

dd$SNP_1.1 == dd$SNP_1.2

```
</div>

This gives us `r nrow(dd)` `TRUE` or `FALSE` values, and tells us whether each individual in our dataset is a homozygote for SNP1 or not. It might be more helpful to have a vector of values, stating explicitly whether each individual is a homozygote or a heterozygote. We can create this using the `ifelse()` function. Check out what this function does using `?ifelse`. Can you generate a column saying whether each individual in your dataset is a homozygote or heterozygote for the SNP1? You should end up with something like this:

<div class="fold s">
```{r}

#Add a Heterozygosity column
dd$SNP1_Het <- ifelse(dd$SNP_1.1 == dd$SNP_1.2,"Homozygote","Heterozygote")

#Check that it's worked
dd$SNP1_Het

```

</div>

We can see by checking back to our original data that the homozygotes and heterzygotes are being correctly identified. Now we can do some calculations and make a plot for this first locus.

###Step three - calculate and plot genotype frequencies

Now we want to calculate the proportion of SNP1 heterozygotes in each of our two populations. To do this, we need to know i) the number of heterozygotes in each population and ii) the total number of individuals in each population. We can get both of these pieces of information using the function `count()` from the `dplyr` package.

```{r,fig.width = 4}
count(dd,SNP1_Het)
count(dd,Pop)
count(dd,SNP1_Het,Pop)


```

Note that this function uses the name of the tibble *dd* as the first argument, then you specify which columns you want to count. This is a feature of almost every function in the tidyverse, and we'll get to why it's important later. 

We can assign the output of the `count()` function to an object. Then using this we can extract subsets of data and perform calculations using them. Using this approach, we can divide the number of heterozygotes by the sample size to calculate observed heterozygosity for SNP1. 

```{r}
#Get the genotypes for each population
geno_counts <- count(dd,SNP1_Het,Pop)
geno_counts

#Use filter to pull out the heterozygote rows
het_counts <- filter(geno_counts, SNP1_Het == "Heterozygote")
het_counts

#Get total number of individuals per population
sample_sizes <- count(dd, Pop)
sample_sizes

#Now divide the two together and add them to a new column
het_counts$het_freqs <- het_counts$n/sample_sizes$n
het_counts
```


Now we can plot the frequencies. We will get into plotting in much more detail later in the course. For now, we will use the function `barplot()`.

```{r fig.width = 4}
barplot(height = het_counts$het_freqs, names.arg = het_counts$Pop,ylab = "Proportion heterozygotes")
```

We can test whether the proportion of heterozygotes differs between mainland and island population using a Chi-squared test (`?chisq.test`).

```{r}
chisq.test(dd$SNP1_Het,dd$Pop)
```

What are you able to conclude from this? Is the effect significant? Is it in the expected direction? What is lacking from this data?

Finally, we may wish to write our summarised data to a file. We can do this using the functions `write_delim()` (to write a tab separated table), or `write_csv` to write a CSV file. Let's write our geneotype counts to file now. Note that you will have to change the file path depending on where you want to write your file.

```{r}

write_delim(geno_counts,path = "~/Documents/Teaching/PopGenBerlinLGS/example_genotype_counts.txt")
write_csv(geno_counts,path = "~/Documents/Teaching/PopGenBerlinLGS/example_genotype_counts.csv")

```

Open your text and csv files in sublime and/or excel. How do they look? Any problems? Can you use the help files to figure out how to remove the quotes in the tab separated text file?


##If you've finished

- Can you calculate heterozygosity for all five SNPs? Can you find a way to do it more quickly and efficiently than repeating the same code?
- Following this, can you make a plot something like the below? You can compare your code to mine, which is in the script "Mainland_Island_SNP_analysis.R"


```{r echo = F,fig.width = 6,fig.height=4,fig.align="center",message = F}

source("Mainland_Island_SNP_analysis.R")
p

```


- Calculate and plot the frequency distribution of minor alleles in each population.

- Calculate expected genotype frequencies for each SNP and population, according to Hardy-Weinberg. Compare your observed and expected frequencies using chi-squared tests.

- I simulated the dataset for this practical in R. However, I did a dreadful job of annotating the script that I used to generate this dataset. Go through the script "Mainland_Island_SNP_data.R", figure out what is being done with each line of code, and annotate the script accordingly.

- Try simulating your own SNP dataset by developing my code, or your own.