---
title: 'Genetic diversity: practical'
author: "Lewis Spurgin"
output:
  html_document:
    css: js/styles.css
---

<script src="js/hide.js"></script>

## Introduction
The aim of this practical is to get you estimating genomic diversity from large SNP data files. We will first become comfortable reading and subsetting very large data files. We will then use VCFtools to estimate some common measures of genetic diversity in individuals and populations. We will then develop our skills in R to manipulate our output files from plink and VCFtools. Finally, we will introduce ggplot2, and learn how to use this package to produce publication quality figures of our data.

Once again, you should try to figure things out for yourself before clicking on "show". To do this, you will need the documentation for VCFtools open. It can be found at https://vcftools.github.io/man_latest.html

## The data
We will be working with some of my own genomic data obtained from experimentally evolved populations the red flour beetle *Tribolium castaneum*. The data file includes SNP information (in vcf format) from six resequenced genomes. Three of these individuals come from a population that has been experimentally evolved under male-biased sex ratios for ~80 generations, while the other three come from a population with a female-biased sex ratio (see Lumley et al. 2015, *Nature* for details of the experimental lines). We wish to know whether levels of genomic diversity differs between these two populations as a result of differing patterns of sexual selection.

Log onto the Amazon EC2 server and change your directory to the `genetic_diversity` folder within your own set of files. You should see two files. `TriboliumSSGenomes.vcf` and  `Tribolium_sample_names.csv`. We can't open these files in the server, but we can explore their properties and look at them using the command line. Try to do the following:

1. Find out the size of each of the two files
2. Print the contents of `Tribolium_sample_names.csv` to the terminal
3. Print the first 11 lines of `TriboliumSSGenomes.vcf` to the terminal

<div class="fold s">
```{r, eval = F}
ls -l * #Get info for all file in directory, including file size - could also use wc (word count)
cat Tribolium_sample_names.csv #Print file to console
head -n 11 TriboliumSSGenomes.vcf #Get first 100 lines

```
</div>

These files are in more or less exactly the same format as when received from a sequencing facility. The facility used their own sample names, and the file `Tribolium_sample_names.csv` lists our original sample names followed by the sample name used in the vcf file. This will become important later. Note for now that the sample names are coded with population and individual ID ("MA" for the male-biased population and "FB"" for the female-biased population, followed by "_", followed by the individual ID).

Before calculating heterozygosity, we will first check that the file can be read into the program. Try reading in the data into VCFtools without performing any additional operations.

<div class="fold s">
```{r, eval = F}
vcftools --vcf TriboliumSSGenomes.vcf
```
</div>

What does VCFtools give you? What do you now know about the dataset?

The two most basic ways that we can estimate genetic diversity using vcftools are i) by estimating the number of heterozygous sites (and an inbreeding coefficient) for each individual, and ii) by estimating observed and expected heterozygosity for each marker. Can you find the command for each of these and run them on the "TriboliumSSGenomes.vcf" dataset. Remember to specify a suitable output name.

<div class="fold s">
```{r, eval = F}
vcftools --vcf TriboliumSSGenomes.vcf --het --out TriboliumSSGenomes
vcftools --vcf TriboliumSSGenomes.vcf --hardy --out TriboliumSSGenomes
```
</div>

Now use `ls -l`, `head`, `cat`,`less` etc., as appropriate, to explore the output files from each of these analyses. What have the analyses done? Have you noticed any problems with the way the analyses have been performed?

Hopefully you did noticed that the `.hwe` file was very large and didn't try to use `cat` to look at it! Using `head` we can see that the `--hardy` analysis has calculated observed and expected heterozygosity for every SNP. However, we it has averaged over our two populations for this. **Before reading on**, try to figure out how you would perform this analysis separately for each population. Then discuss with a neighbour whether you think this analysis is appropriate for this dataset (clue - think sample sizes), and think of other appropriate analyses.

We are now going to perform some analyses of nucleotide diversity. If you look at the vcftools documentation you will see that there are two options for this. We can either calculate nucleotide diversity on a "per site" basis, or in "windows" of a defined size. With just three individuals per population, per-site based measures of diversity (including the `--hardy` analysis performed just now) are problematic as each site is extremely limited in the possible values that can be obtained. We can get around this using windows. Discuss with a neighbour what is meant by "windows" and why they might be preferable to per-site measured of diversity in this instance.

We would like to calculate diversity in non-overlapping 200kb windows for each of our two populations. Can you figure out how to do this in vcftools?

For a hint, see the box below.
<div class="fold s">
```{r, eval = F}
#You will need the following commands, and to look at the Tribolium_sample_names.csv file
--window-pi
--indv
```
</div>


For the answer, see here:
<div class="fold s">
```{r, eval = F}
#For the male-biased population
vcftools --vcf TriboliumSSGenomes.vcf --window-pi 200000 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB

#For the female-biased population
vcftools --vcf TriboliumSSGenomes.vcf --window-pi 200000 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA

```
</div>

Hopefully you figured out that individuals 2014_SOL_MG4955, 2014_SOL_MG4956 and 2014_SOL_MG4957 are from the MB population, while the remaining individuals are from the FA population. Once you have run the above code, use the command line to take a look at the format of the output files vcftools has returned. Does it makes sense?

Before we have a go at plotting our data, we will first calculate one more measure of genetic diversity - linkage disequilibrium. Our aim is to look at how LD decays with physical distance between SNPs, and to see whether this relationship varies between our two populations. Lower levels of LD and faster rates of LD decay are typically found in large, outbred populations, whereas in small bottlenecked populations background levels of LD are higher and LD decays much more slowly.

Because LD is calculated between *pairs* of markers, the number of calculations VCFtools would have to perform for a whole genome dataset would be huge, and unnecessary for our purposes here. We will therefore restrict our analyses to i) a single chromosome (ChLG1), ii) pairs of markers within 100 kb of each other, and iii) pairs of markers with R2 values > 0.001. Using this information, can you calculate LD for both the MB and FA populations? Answer below.

<div class="fold s">
```{r eval = F}
vcftools --vcf TriboliumSSGenomes.vcf --geno-r2 --chr ChLG1 --ld-window-bp 100000 --min-r2 0.001 --indv 2014_SOL_MG4955 --indv 2014_SOL_MG4956 --indv 2014_SOL_MG4957 --out TriboliumSSGenomesMB

vcftools --vcf TriboliumSSGenomes.vcf --geno-r2 --chr ChLG1 --ld-window-bp 100000 --min-r2 0.001 --indv 2014_SOL_MG4972 --indv 2014_SOL_MG4973 --indv 2014_SOL_MG4974 --out TriboliumSSGenomesFA

```
</div>

Use `ls - l` to look at the size of the output files. You will see that they are *huge*. Even with our restrictions, we have more than 50 million rows in each of our files. So instead of downloading this onto our laptops for plotting, we will perform a bit of extra processing on the data in the server, using R. Open up a text editor (or R studio) and save an R script using a suitable name. Then, on the Amazon server, start up R. We are going to type our instructions into a script, so we have a record of what we've done, but at the same time we will use the console interactively. First, we need to load a library. At the top of your script, add the following code, then copy and paste the command into your console:

```{r eval = F}
R
library(data.table)
```

The library `data.table` includes the very useful function `fread()`, which is designed to read in very large text files. Take a look at the help file for this function. Now use `fread()` to read in the two `.geno.ld` files you have created (one for each population), and assign each data frame to a different object name.

<div class="fold s">
```{r eval = F}
MB <- fread("TriboliumSSGenomesMB.geno.ld")
FA <- fread("TriboliumSSGenomesFA.geno.ld")
```
</div>

Have a look at what's contained in these data files using `head()` and/or `str()`. You will see that they are very large indeed. Each row contains the LD statistic and other information for a pair of SNPs. There are five columns: the chromosome that the two SNPs are on, the positions (in bp) of the two SNPs, the number of individuals tested, and the R2 value. We want to plot how R2 declines with the physical distance between pairs of SNPs. However, we can't plot all of our data as i) there's too much, and ii) with three individuals the R2 values are two noisy. So as before, we are going to use windowed averages. This time we will use 1000bp windows, as the maximum distance between any pair of SNPs in our data is 100,000. Our aim is to reduce our two data files each containing millions of LD measures to a single data file with 100 averaged LD measures for each population.

The first thing we wish to do is calculate the physical distance between each pair of SNPs. This can be obtained by subtracting the SNP positions.

<div class="fold s">
```{r eval = F}
MB$dist <- MB$POS2 - MB$POS1
FA$dist <- FA$POS2 - FA$POS1

```
</div>

However, we do not actually want to know the exact distance between each pair of SNPs, but rather the distance rounded *up* to the nearest 1000 bp. Rather than create an extra variable, we can modify our previous command to obtain the distance between each pair of SNPs rounded up to the nearest 1kb.

For a hint of how to do this click here:
<div class="fold s">
```{r eval = F}
x <- 1550
ceiling(x/1000)*1000
```
</div>

For an answer click here:
<div class="fold s">
```{r eval = F}
MB$dist <- ceiling((MB$POS2 - MB$POS1)/1000)*1000
FA$dist <- ceiling((FA$POS2 - FA$POS1)/1000)*1000
```
</div>

Now we need to obtain the mean R2 value for each 1000bp distance bin. For this we are going to use the `ddply()` function in the package `plyr`. With  `ddply()`, we can perform functions on variables data frames (e.g. calculate means), and return the output in a data frame. Have a look at the relevant help pages and try searching for help online. Here is a tutorial:

http://www.cookbook-r.com/Manipulating_data/Summarizing_data/

If you think you understand how it works, have a go at using `ddply()` to create a new data frame with averaged R2 values for each 1000bp window.


<div class="fold s">
```{r eval = F}
library(plyr)

MB2 <- ddply(MB,
      .(dist),
      summarise,
      meanR2 = mean(`R^2`))

FA2 <- ddply(FA,
      .(dist),
      summarise,
      meanR2 = mean(`R^2`))

#Combine
dd <- rbind(MB2,FA2)

#Add a population column
dd$pop <- rep(c("MB","FA"),c(nrow(MB2),nrow(FA2)))

write.csv(dd,"TriboliumSSGenomes.windowed.ld.csv",row.names = F)

```
</div>

Have a look at one of your resulting outputs. Does it all make sense?

That's our analyses done! Now you can log out of the server, and use SCP to download the genetic results that you have just created. To recap, you should download the files for:

- individual heterozygosity (one file, one value per individual), 
- windowed nucleotide diversity (two files - one per population), 
- windowed LD (one file). 

Make a new folder called "session_3_output" or something similar using `mkdir` Download your four genetic diversity outputs using `scp` (mac) or `pscp` (windows), in exactly the same way as you learned this morning. Once you have done this, open R studio on your laptop and use `read.csv()` and `read.table()` as appropriate to read in your newly created and downloaded data files to four suitably-named objects.

<div class="fold s">
```{r}
rm(list=ls()) #Clear your workspace

het <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session_3_output/TriboliumSSGenomes.het",header = T)
pi_MB <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session_3_output/TriboliumSSGenomesMB.windowed.pi", header = T)
pi_FA <- read.table("~/Documents/Teaching/PopGenBerlinLGS/session_3_output/TriboliumSSGenomesFA.windowed.pi", header = T)
ld <- read.csv("~/Documents/Teaching/PopGenBerlinLGS/session_3_output/TriboliumSSGenomes.windowed.ld.csv") #Or whatever you have called it

```
</div>

We are going to use the `ggplot2` library to make the following plot. If you think you can recreate this, please give it a shot. Otherwise, we will work through the plots step-by-step below.

```{r echo = F,warning = F,message=F,fig.height=6}
library(Rmisc)
library(plyr)
library(magrittr)
library(ggplot2)

Fig1A <- mutate(het,Population = c(rep("MB",3),rep("FA",3)),Heterozygosity = 1-(O.HOM./N_SITES)) %>%
  ggplot(aes(x = Population, y = Heterozygosity,col = Population))+
  geom_point()+
  theme_bw()+
  theme(legend.position = "none")+
  xlab("")+
  ggtitle("A")

Fig1B <- rbind(pi_MB,pi_FA) %>%
  mutate(Population = c(rep("MB",nrow(pi_MB)),rep("FA",nrow(pi_FA)))) %>%
    ggplot(aes(x = Population, y = PI,col = Population))+
  geom_jitter(col = "grey",width = 0.1)+
  geom_boxplot(notch = T,alpha = 0,lwd = 0.5,outlier.shape = NA)+
  theme_bw()+
  theme(legend.position = "none")+
  xlab("")+
  ylab(expression(pi))+
  ggtitle("B")

Fig1C <-  mutate(ld,Population = pop,distkb = dist/1000) %>%   
  ggplot(aes(x = distkb, y = meanR2,col = Population))+
  geom_point()+
  geom_line()+
  theme_bw()+
  xlab("Distance (kb)")+
  ylab(expression("R"^2))+
  ggtitle("C")

multiplot(Fig1A,Fig1B,Fig1C,layout = matrix(c(1,2,3,3),nrow = 2, byrow = T))
```
<br>
**Figure 1** Genomic variation in two experimentally evolved red flour beetle populations. **A** Observed heterozygosity (proportion of heterozygous sites) in each individual. **B** Nucleotide diversity in each population, estimated from 200kb non-overlapping windows. **C** Linkage disequilibrium between marker pairs the first 1mb of chromosome 6. R^2^ values are means of marker-pairs within 1kb non-overlapping bins.

<br><br>

We will work through the figures one at a time. We will start with panel **A**. Let's first take a look at our data:

```{r}
head(het)
```

As we have already seen, the `--het` function in vcftools gives us, for each individual, the number of homozygous sites, the number of heterozygous sites, the total number of sites and the inbreeding coefficient F. In panel **A** you can see that I have plotted the *proportion* of heterozygous sites per individual. Can you calculate this in R, and assign individual heterozygosity to a new column? You will also notice that there is no "population" column, and we will need one. Can you use the the available information (hint: see previous code) to assign a population column?

<div class="fold s">
```{r}
het$Heterozygosity <- 1-(het$O.HOM./het$N_SITES)
het$Population <- c(rep("MB",3),rep("FA",3)) #Know from previously that the first three individuals are from MB
```
</div>

Take a look at the structure of het. How has the population column been stored? Can you turn it into a factor?

<div class="fold s">
```{r}
het$Population <- factor(het$Population)
```
</div>



We can try plotiing heterozygosity against population using the base plotting function in R. We want heterozygosity on the y axis, and population ID on the x. So we would plot as follows:

```{r}
plot(Heterozygosity~Population,data = het)
```

Base graphics automatically generates a boxplot, which is not appropriate for our data (as we only have three points per population!). We could get around this, but it is a lot of unnecessary coding and work. 

Instead, for the rest of this course we will be using ggplot2 for our plots - if you have no experience using this package, take 5-10 minutes to familiarise yourself by searching for an online tutorial. Here is one example:

http://www.noamross.net/blog/2012/10/5/ggplot-introduction.html


Basically, `ggplot()` builds up a plot in layers. We first tell R *what* we want to plot, then we tell it *how* to plot it (e.g. with points, boxplot etc.) 

We can use ggplot2 to plot heterozygosity against population. We will start off defining the aesthetics, which produces a blank plot. by using the function `geom_point()`, as follows:

```{r}
#Load the ggplot library
library(ggplot2)

#Make a plot
A <- ggplot(het,aes(x = Population, y = Heterozygosity))

#Call the plot up
A
```


Can you figure out how to now tell ggplot to add points to the plot?

<div class="fold s">
```{r}
A <- ggplot(het,aes(x = Population, y = Heterozygosity)) +
  geom_point()

A
```
</div>

This is already pretty close to what we are after. We just need to colour the points by population:

<div class="fold s">
```{r}
A <- ggplot(het,aes(x = Population, y = Heterozygosity, col = Population)) +
  geom_point()

A
```
</div>

Now we need to change the background colour. I did this by adding a layer called `theme_bw()`. Check out ggplots "themes" online for more:

<div class="fold s">
```{r}
A <- ggplot(het,aes(x = Population, y = Heterozygosity, col = Population)) +
  geom_point()+
  theme_bw()

A
```
</div>

Remove the x axis label and the legend:

<div class="fold s">
```{r}
A <- ggplot(het,aes(x = Population, y = Heterozygosity, col = Population)) +
  geom_point()+
  theme_bw()+
  theme(legend.position = "none")+
  xlab("")

A
```
</div>

Add the panel label:

<div class="fold s">
```{r}
A <- ggplot(het,aes(x = Population, y = Heterozygosity, col = Population)) +
  geom_point()+
  theme_bw()+
  theme(legend.position = "none")+
  xlab("")+
  ggtitle("A")

A
```
</div>

And we're there! Now let's move onto panel B, which is slightly more complicated. Again, we first look at the data. Remember we have two data frames for the nucleotide diversity data - one for each population:

```{r}
str(pi_MB)
str(pi_FA)
```

Because ggplot likes to work from a single data frame, we first need to combine our data from the two populations and add a population ID column:

```{r}
pi_both <- rbind(pi_MB,pi_FA)
pi_both$Population <- c(rep("MB",nrow(pi_MB)),rep("FA",nrow(pi_FA)))
```

Now we are ready to make a plot. We will start with the box plot. Remember that we can use a lot of the same arguments from our previous plot. But this time we will use `geom_boxplot()` instead of `geom_point()`.

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population)) +
      geom_boxplot()+
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")
B
```
</div>

We can add a notch to the box plot and remove the outliers (google!):

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population)) +
      geom_boxplot(notch = T,outlier.shape = NA)+
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")
B
```
</div>


To add the points behind the box plot we will use the function `geom_jitter`:

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population)) +
      geom_boxplot(notch = T,outlier.shape = NA)+
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")+
      geom_jitter()
B
```
</div>

There are two problems here. First, the points are getting in the way of the box plots. Second, they have been coloured by population whereas we would like them to be grey. Any idea how we fix these things?

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population))+
      geom_jitter(col = "grey")+ #Add this before the geom_boxplot() to put the points behind the boxes
      geom_boxplot(notch = T, alpha = 0,outlier.shape = NA)+ #alpha makes the background seethrough
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")

B
```
</div>

Getting there... the points in this plot are spread more widely here than in the plot above though. Do you know how we can pack them in more?

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population))+
      geom_jitter(col = "grey",width = 0.1)+ #Add this before the geom_boxplot() to put the points behind the boxes
      geom_boxplot(notch = T, alpha = 0,outlier.shape = NA)+ #alpha makes the background seethrough
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")

B
```
</div>

Finally, we just need to add the panel label and sort out our y axis. How can we add a symbol? (Hint - look up `expression()`):

<div class="fold s">
```{r}
B <- ggplot(pi_both,aes(x = Population, y = PI, col = Population))+
      geom_jitter(col = "grey",width = 0.1)+ #Add this before the geom_boxplot() to put the points behind the boxes
      geom_boxplot(notch = T, alpha = 0,outlier.shape = NA)+ #alpha makes the background seethrough
      theme_bw()+
      theme(legend.position = "none")+
      xlab("")+
      ylab(expression(pi))+
      ggtitle("B")
B
```
</div>

Done! Now we need to make the LD plot and stick them all together. Let's take another look at the LD data:
```{r}
head(ld)
```
This should be relatively simple by now, as we can use the other two plots as templates, and our data is already conveniently formatted in the right way. Let's start by plotting the points, with distance between marker pairs on the x axis, R2 on the y axis, with colours corresponding to populations.

<div class="fold s">
```{r}
C <- ggplot(ld,aes(x = dist, y = meanR2, col = pop)) +
      geom_point()+
      theme_bw()
C
```
</div>

Notice that distance is in bp, whereas we would like it in kb. We can also add the lines:

<div class="fold s">
```{r}
C <- ggplot(ld,aes(x = dist/1000, y = meanR2, col = pop)) +
      geom_point()+
      geom_line()+
      theme_bw()
C
```
</div>

All we need to do now is add the panel label, sort out the x axis and y axis labels (hint: you will need `expression()` again for the y axis), and change the legend title. See if you can figure out how to do the the latter on your own (there are multiple ways).

<div class="fold s">
```{r}
C <- ggplot(ld,aes(x = dist/1000, y = meanR2, col = pop)) +
      geom_point()+
      geom_line()+
      theme_bw()+
      xlab("Distance (kb)")+
      ylab(expression(R^2))+
      scale_colour_discrete(name = "Population")+ #An alternative would be to just rename to pop column to Population.
      ggtitle("C")
C
```
</div>

Ok, we have done our plots! We now need to make a single, muti-panel plot. Again, there are lots of ways we can do this and this is definitely something you should explore. For now we will use the `multiplot()` function from the `Rmisc` library. We can just list the plots without specifying anything to do with the layout:

```{r}
library(Rmisc)
multiplot(A,B,C)
```
<br>
However, this stacks the plots one on top of the other. We can alter the layout in an enormous number of ways using the argument `layout` and the function `matrix()`. Can you figure out what is going on here? If not, try looking up the `layout()` function and seeing how this works.

```{r,fig.height=6}
multiplot(A,B,C,layout = matrix(c(1,2,3,3),nrow = 2,byrow = T))
```

##If you've finished
- Use `multiplot()` to crease a multi-panelled plot with i) the LD plot on the top, and ii) the heterozygosity plot stretched along the left-hand side, with the pi and LD plots both on the right hand side.
- Plot and test whether nucleotide diversity in each 200kb window is correlated between the two populations. What might explain such a pattern?
- Recreate this plot of nucleotide diversity (200kb windows) for each chromosome, as precisely as you can: 

```{r, echo = F,fig.width = 9,fig.height = 5}

pi_both$CHROM <- factor(pi_both$CHROM,levels = paste0("ChLG",c(1:10)))

X <- ggplot(pi_both,aes(x = CHROM, y = PI, col = Population,fill = Population))+
      geom_jitter(position = position_jitterdodge(jitter.width = 0.1))+ #Add this before the geom_boxplot() to put the points behind the boxes
      geom_boxplot(notch = T, alpha = 0.2,col = "black",outlier.shape = NA)+ #alpha makes the background seethrough
      theme_bw()+
      xlab("")+
      ylab(expression(pi))
X
```
